File: content/snippets/go/interview_lcci/08/10_color_fill.go
package main

import "fmt"

func floodFill(image [][]int, sr int, sc int, newColor int) [][]int {
	old := image[sr][sc]
	if old == newColor {
		return image
	}
	r, c := len(image), len(image[0])

	var dfs func(i, j int)
	dfs = func(i, j int) {
		if i < 0 || i >= r || j < 0 || j >= c || image[i][j] != old {
			return
		}
		image[i][j] = newColor
		dfs(i+1, j)
		dfs(i-1, j)
		dfs(i, j+1)
		dfs(i, j-1)
	}

	dfs(sr, sc)
	return image
}

func main() {
	image := [][]int{{1, 1, 1}, {1, 1, 0}, {1, 0, 1}}
	result := floodFill(image, 1, 1, 2)
	for _, row := range result {
		fmt.Println(row)
	}
}

File: content/snippets/cpp/interview_lcci/08/10_color_fill.cpp
#include <iostream>
#include <vector>
#include <functional>
using namespace std;

vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
    int old = image[sr][sc];
    if (old == newColor) return image;
    int r = image.size(), c = image[0].size();
    function<void(int, int)> dfs = [&](int i, int j) {
        if (i < 0 || i >= r || j < 0 || j >= c || image[i][j] != old) return;
        image[i][j] = newColor;
        dfs(i + 1, j); dfs(i - 1, j); dfs(i, j + 1); dfs(i, j - 1);
    };
    dfs(sr, sc);
    return image;
}

int main() {
    vector<vector<int>> image = {{1,1,1},{1,1,0},{1,0,1}};
    auto result = floodFill(image, 1, 1, 2);
    for (auto& row : result) {
        for (int i = 0; i < (int)row.size(); i++) {
            cout << row[i] << (i + 1 < (int)row.size() ? " " : "\n");
        }
    }
    return 0;
}
